prow/plugins/golint/suggestion/golint_suggestion.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package suggestion
+
+import (
+	"github.com/golang/lint"
+	"regexp"
+	"strings"
+)
+
+var (
+	lintNamesRegex1 = regexp.MustCompile("don't use underscores in Go names; (.*) should be (.*)")
+	lintNamesRegex2 = regexp.MustCompile("don't use ALL_CAPS in Go names; use CamelCase")
+)
+
+var allCapsRE = regexp.MustCompile(`.+[A-Z0-9_]+.+$`)
+
+var lintCaseMap = map[*regexp.Regexp]func(*regexp.Regexp, lint.Problem, []string) string{
+	lintNamesRegex1: fixNameUnderscore,
+	lintNamesRegex2: fixNameAllCaps,
+}
+
+func getMatches(p lint.Problem) []string {
+	for r := range lintCaseMap {
+		matches := r.FindStringSubmatch(p.Text)
+		if len(matches) > 0 {
+			return matches
+		}
+	}
+	var ret []string
+	return ret
+}
+
+// SuggestCodeChange returns code suggestions for a given lint.Problem
+// Returns empty string if no suggestion can be given
+func SuggestCodeChange(p lint.Problem) string {
+	matches := getMatches(p)
+	if len(matches) == 0 {
+		return ""
+	}
+	var matchedRegex *regexp.Regexp
+	var codeSuggestionFunc func(*regexp.Regexp, lint.Problem, []string) string
+	for r, c := range lintCaseMap {
+		result := r.MatchString(p.Text)
+		if result {
+			codeSuggestionFunc = c
+			matchedRegex = r
+		}
+	}
+	if matchedRegex == nil {
+		return ""
+	}
+
+	var result = codeSuggestionFunc(matchedRegex, p, matches)
+	if result == p.LineText {
+		return ""
+	}
+	return formatSuggestion(result)
+}
+
+func fixNameUnderscore(re *regexp.Regexp, p lint.Problem, matches []string) string {
+	name := matches[1]
+	words := strings.Fields(name)
```

Comment:
I needed to split `name` into fields because `matches[1]` has two words: `func Qux_Func`, but thanks for your suggestion I'm able to simple this much further now with        

```
underscoreRe := regexp.MustCompile(`[A-Za-z]+(_[A-Za-z0-9]+)+`)
namesWithUnderscore := underscoreRe.FindStringSubmatch(matches[1])
suggestion := strings.Replace(p.LineText, namesWithUnderscore[0], matches[2], -1)
```